File: backend/app/controllers/companies.py
from uuid import UUID

from fastapi import Depends, HTTPException
from fastapi.responses import Response
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.db import get_db
from app.models import Company, CompanyImportBatch
from app.services import company_directory, excel_pipeline, llm_client, signal_directory
from app.schemas.company import (
    CompanyInsightOut,
    CompanyListItemOut,
    CompanyListOut,
    CompanyStatsOut,
    CountryLeadScoreOut,
)

XLSX_MEDIA_TYPE = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"


async def list_companies(
    organisation_id: UUID,
    page: int = 1,
    page_size: int = 25,
    search: str | None = None,
    import_batch_id: UUID | None = None,
    db: AsyncSession = Depends(get_db),
):
    page_size = min(page_size, 100)
    rows, total = await company_directory.list_companies(
        db, organisation_id, page, page_size, search, import_batch_id
    )
    def _num(v):
        return float(v) if v is not None else None

    items = [
        CompanyListItemOut(
            company_id=company.company_id,
            company_name=company.company_name,
            company_domain=company.company_domain,
            city=company.city,
            state=company.state,
            country=company.country,
            employee_count=company.employee_count,
            employee_range=company.employee_range,
            revenue_usd=company.revenue_usd,
            revenue_range=company.revenue_range,
            industries=company.industries,
            logo_url=company.logo_url,
            lead_score=_num(ls.lead_score) if ls else None,
            sales_status=ls.sales_status if ls else None,
            confidence_label=ls.confidence_label if ls else None,
            buying_evidence_score=_num(ls.buying_evidence_score) if ls else None,
            contact_access_score=_num(ls.contact_access_score) if ls else None,
            negative_event_score=_num(ls.negative_event_score) if ls else None,
            best_offering=ls.best_offering if ls else None,
            why_now=ls.why_now if ls else None,
            expected_deal_value_usd=_num(ls.expected_deal_value_usd) if ls else None,
        )
        for company, ls in rows
    ]
    return CompanyListOut(items=items, total=total, page=page, page_size=page_size)


async def stats(organisation_id: UUID, import_batch_id: UUID | None = None, db: AsyncSession = Depends(get_db)):
    """Evidence-based company stats (brief items 22, 4): sales-status bands +
    confidence + provisional pipeline value, optionally batch-scoped."""
    summary = await company_directory.sales_status_summary(db, organisation_id, import_batch_id)
    country_rows = await company_directory.lead_score_by_country(db, organisation_id, import_batch_id)
    return CompanyStatsOut(
        total=summary["total"],
        scored=summary["scored"],
        unscored=summary["unscored"],
        sales_ready=summary["sales_ready"],
        high_priority=summary["high_priority"],
        warm=summary["warm"],
        monitor=summary["monitor"],
        low_priority=summary["low_priority"],
        high_confidence=summary["high_confidence"],
        provisional_pipeline_value=summary["pipeline_value"],
        by_country=[
            CountryLeadScoreOut(country=country, avg_lead_score=float(avg or 0), company_count=count)
            for country, avg, count in country_rows
        ],
    )


async def insight(organisation_id: UUID, db: AsyncSession = Depends(get_db)):
    """Dashboard briefing over the evidence-based pipeline (brief section 29).
    No gates, no ICP - framed around Buying Evidence + Contact Access -
    Negative Evidence, sales-status bands, and expected pipeline value."""
    summary_data = await company_directory.sales_status_summary(db, organisation_id)
    total = summary_data["total_scored"]
    if total == 0:
        return CompanyInsightOut(summary="No scored companies yet - upload prospect data to begin research and scoring.")

    rows, _ = await company_directory.list_companies(db, organisation_id, page=1, page_size=5)
    top_companies = [
        {
            "name": company.company_name,
            "lead_score": round(float(ls.lead_score), 1),
            "sales_status": ls.sales_status,
            "confidence": ls.confidence_label,
            "best_offering": ls.best_offering,
            "why_now": ls.why_now,
            "expected_deal_value_usd": float(ls.expected_deal_value_usd) if ls.expected_deal_value_usd is not None else None,
        }
        for company, ls in rows
        if ls is not None and ls.lead_score is not None
    ]

    avg_score_str = f"{summary_data['avg_lead_score']:.1f}" if summary_data["avg_lead_score"] is not None else "not yet available"
    pipeline_value = summary_data["pipeline_value"]

    prompt = (
        "You are a senior B2B sales intelligence analyst for XSparks (an AI solutions partner), writing a "
        "daily pipeline briefing. Scoring is evidence-based: Lead Score = Buying Evidence + Contact Access "
        "- Negative Evidence (0-100); revenue/funding affect only Expected Deal Value, never the Lead "
        "Score; confidence is separate from score. Write 3 short plain-prose paragraphs (no headings, no "
        "markdown, no bullets):\n\n"
        "Paragraph 1 - Pipeline health: total scored companies, the sales-status split (Sales Ready / High "
        "Priority / Warm / Monitor / Low Priority), the average lead score, and the estimated pipeline "
        "value.\n"
        "Paragraph 2 - Top opportunities: name the specific top-scored companies, their lead scores, the "
        "best XSparks offering and why-now for each - interpret for a sales leader, don't just repeat "
        "numbers.\n"
        "Paragraph 3 - Recommended action: one concrete recommendation for what the team should prioritise "
        "today, grounded in the evidence.\n\n"
        "Never invent a number, company, or fact not given below. Confident, professional analyst tone.\n\n"
        f"DATA:\n"
        f"- Scored companies: {total}. Sales Ready {summary_data['sales_ready']}, High Priority "
        f"{summary_data['high_priority']}, Warm {summary_data['warm']}, Monitor {summary_data['monitor']}, "
        f"Low Priority {summary_data['low_priority']}.\n"
        f"- Average lead score: {avg_score_str} / 100.\n"
        f"- Estimated pipeline value (sum of expected deal values): ${pipeline_value:,.0f}.\n"
        f"- Top companies: {top_companies}."
    )

    try:
        summary = await llm_client.complete(
            [{"role": "user", "content": prompt}],
            generation_name="dashboard-company-overview",
            trace_user_id=str(organisation_id),
        )
    except llm_client.LLMNotConfiguredError:
        top_names = ", ".join(f"{c['name']} ({c['lead_score']})" for c in top_companies[:3]) or "none scored yet"
        summary = (
            f"{total} companies scored - {summary_data['sales_ready']} Sales Ready, "
            f"{summary_data['high_priority']} High Priority, {summary_data['warm']} Warm, "
            f"{summary_data['monitor']} Monitor, {summary_data['low_priority']} Low Priority. "
            f"Average lead score {avg_score_str}/100. Estimated pipeline value ${pipeline_value:,.0f}. "
            f"Top scored: {top_names}."
        )

    return CompanyInsightOut(summary=summary)


async def export(organisation_id: UUID, import_batch_id: UUID | None = None, db: AsyncSession = Depends(get_db)):
    """Evidence-based company export (brief section 23). Optionally scoped to a
    single upload via import_batch_id; no ICP."""
    company_ids = None
    if import_batch_id is not None:
        # Permanent membership table (item 5), not Company.import_batch_id -
        # a later re-upload of the same company must not exclude it from an
        # earlier batch's export.
        company_ids = set(
            (
                await db.execute(
                    select(Company.company_id).where(
                        Company.organisation_id == organisation_id,
                        Company.company_id.in_(
                            select(CompanyImportBatch.company_id).where(
                                CompanyImportBatch.import_batch_id == import_batch_id
                            )
                        ),
                    )
                )
            ).scalars().all()
        )

    rows = await company_directory.list_companies_for_export(db, organisation_id, company_ids)
    workbook_bytes = excel_pipeline.build_company_export_workbook(rows)

    return Response(
        content=workbook_bytes,
        media_type=XLSX_MEDIA_TYPE,
        headers={"Content-Disposition": 'attachment; filename="companies_export.xlsx"'},
    )


async def get_company(organisation_id: UUID, company_id: UUID, db: AsyncSession = Depends(get_db)):
    company = await company_directory.get_company(db, organisation_id, company_id)
    if company is None:
        raise HTTPException(status_code=404, detail="company not found")
    return company


async def list_decision_makers(organisation_id: UUID, company_id: UUID, db: AsyncSession = Depends(get_db)):
    return await company_directory.list_decision_makers(db, organisation_id, company_id)


async def get_decision_maker(
    organisation_id: UUID, decision_maker_id: UUID, db: AsyncSession = Depends(get_db)
):
    dm = await company_directory.get_decision_maker(db, organisation_id, decision_maker_id)
    if dm is None:
        raise HTTPException(status_code=404, detail="decision maker not found")
    return dm

File: frontend/src/api/companies.ts
/* Mirrors backend/app/routes/companies.py */
import { apiGet, apiGetForBlob } from "./client";
import type { CompanyOut, DecisionMakerOut } from "./icp";

export type CompanyListItemOut = {
  company_id: string;
  company_name: string;
  company_domain: string | null;
  city: string | null;
  state: string | null;
  country: string | null;
  employee_count: number | null;
  employee_range: string | null;
  revenue_usd: number | null;
  revenue_range: string | null;
  industries: string[] | null;
  logo_url: string | null;
  lead_score: number | null;
  sales_status: string | null;
  confidence_label: string | null;
  buying_evidence_score: number | null;
  contact_access_score: number | null;
  negative_event_score: number | null;
  best_offering: string | null;
  why_now: string | null;
  expected_deal_value_usd: number | null;
};

export type CompanyListOut = {
  items: CompanyListItemOut[];
  total: number;
  page: number;
  page_size: number;
};

export function listCompanies(
  organisationId: string,
  params: { page?: number; page_size?: number; search?: string; import_batch_id?: string } = {},
): Promise<CompanyListOut> {
  const query = new URLSearchParams();
  if (params.page) query.set("page", String(params.page));
  if (params.page_size) query.set("page_size", String(params.page_size));
  if (params.search) query.set("search", params.search);
  if (params.import_batch_id) query.set("import_batch_id", params.import_batch_id);
  const qs = query.toString();
  return apiGet<CompanyListOut>(`/organisations/${organisationId}/companies${qs ? `?${qs}` : ""}`);
}

export type CountryLeadScoreOut = {
  country: string;
  avg_lead_score: number;
  company_count: number;
  /* The Dashboard globe colors by this, not avg_lead_score - see
   * company_directory.lead_score_by_country: a large mixed population can
   * average into "Monitor" while still holding genuinely Sales Ready
   * companies. */
  max_lead_score: number;
};

export type SectorCountOut = {
  sector: string;
  companies: number;
  scored: number;
  sales_ready: number;
};

export type CompanyStatsOut = {
  total: number;
  scored: number;
  unscored: number;
  sales_ready: number;
  high_priority: number;
  warm: number;
  monitor: number;
  low_priority: number;
  high_confidence: number;
  provisional_pipeline_value: number;
  by_country: CountryLeadScoreOut[];
  /* Industry SECTORS with counts, rolled up server-side from
   * Company.primary_industry (backend/app/core/industry_sectors.py). Sent from
   * the API rather than mapped here so the taxonomy lives in exactly one place -
   * the raw industry column is far too skewed to filter on directly ("Software"
   * alone is ~73% of classified companies). */
  by_sector: SectorCountOut[];
};

/* importBatchId narrows every count to companies from one specific Excel
 * upload (Dashboard timeline picker) instead of everything ever ingested. */
export function getCompanyStats(
  organisationId: string,
  importBatchId?: string,
  sector?: string,
): Promise<CompanyStatsOut> {
  const params = new URLSearchParams();
  if (importBatchId) params.set("import_batch_id", importBatchId);
  /* `sector` narrows by_country (so the globe actually re-colours) but NOT
   * by_sector, so the dropdown keeps listing every sector with its real total
   * instead of collapsing to whichever one is selected. */
  if (sector) params.set("sector", sector);
  const qs = params.toString();
  return apiGet<CompanyStatsOut>(
    `/organisations/${organisationId}/companies/stats${qs ? `?${qs}` : ""}`,
  );
}

export type CompanyInsightOut = {
  summary: string;
};

/* LLM-generated (BridgeLLM, gemini/gemini-2.5-pro) - see
 * backend/app/services/llm_client.py. Falls back to a plain real-numbers
 * sentence server-side if LLM_API_KEY isn't configured, so this never
 * throws just because the key is missing. */
export function getCompanyInsight(organisationId: string): Promise<CompanyInsightOut> {
  return apiGet<CompanyInsightOut>(`/organisations/${organisationId}/companies/insight`);
}

/* Company Directory + evidence-based score columns as an .xlsx download -
 * importBatchId narrows the export to one uploaded batch's companies, omit it
 * to export every company. */
export async function exportCompanies(organisationId: string, importBatchId?: string): Promise<Blob> {
  const qs = importBatchId ? `?import_batch_id=${importBatchId}` : "";
  const { blob } = await apiGetForBlob(`/organisations/${organisationId}/companies/export${qs}`);
  return blob;
}

export function getCompany(organisationId: string, companyId: string): Promise<CompanyOut> {
  return apiGet<CompanyOut>(`/organisations/${organisationId}/companies/${companyId}`);
}

export function listDecisionMakers(organisationId: string, companyId: string): Promise<DecisionMakerOut[]> {
  return apiGet<DecisionMakerOut[]>(`/organisations/${organisationId}/companies/${companyId}/decision-makers`);
}

export function getDecisionMaker(organisationId: string, decisionMakerId: string): Promise<DecisionMakerOut> {
  return apiGet<DecisionMakerOut>(`/organisations/${organisationId}/decision-makers/${decisionMakerId}`);
}
